x/mongo/driver/tranquil_operation.go

```diff
@@ -136,3 +153,374 @@ func (op Operation) decodeOpQuery(wm []byte) (bsoncore.Document, error) {
+		switch op.Legacy {
+		case LegacyListCollections:
+			return op.legacyListCollections(ctx, scratch, srvr, conn, desc)
+		case LegacyListIndexes:
+			return op.legacyListIndexes(ctx, scratch, srvr, conn, desc)
+		}
+	}
+	*/
+
+	var retries int
+	retryable := op.retryable(desc.Server)
+	if retryable && op.RetryMode != nil {
+		switch op.Type {
+		case Write:
+			if op.Client == nil {
+				break
+			}
+			switch *op.RetryMode {
+			case RetryOnce, RetryOncePerCommand:
+				retries = 1
+			case RetryContext:
+				retries = -1
+			}
+
+			op.Client.RetryWrite = false
+			if *op.RetryMode > RetryNone {
+				op.Client.RetryWrite = true
+				if !op.Client.Committing && !op.Client.Aborting {
+					op.Client.IncrementTxnNumber()
+				}
+			}
+		case Read:
+			switch *op.RetryMode {
+			case RetryOnce, RetryOncePerCommand:
+				retries = 1
+			case RetryContext:
+				retries = -1
+			}
+		}
+	}
+	batching := op.Batches.Valid()
+	retryEnabled := op.RetryMode != nil && op.RetryMode.Enabled()
+	return &execSetup{
+		srvr:         srvr,
+		conn:         conn,
+		batching:     batching,
+		desc:         &desc,
+		retryable:    retryable,
+		retryEnabled: retryEnabled,
+		retries:      retries,
+	}, nil
+}
+
+func ReadWireMessageFromConn(ctx context.Context, conn Connection, dst []byte) (hdr *wiremessage.MsgHeader, wm []byte, err error) {
+	wm, err = conn.ReadWireMessage(ctx, dst)
+	if err != nil {
+		return nil, nil, err
+	}
+	var ok bool
+	hdr = &wiremessage.MsgHeader{}
+	hdr.Length, hdr.RequestID, hdr.ResponseTo, hdr.Opcode, _, ok = wiremessage.ReadHeader(wm)
+	if !ok {
+		return nil, nil, errors.New("Incomplete header")
+	}
+	return
+}
+
+func (op Operation) roundTripDirect(ctx context.Context, conn Connection, wm []byte) (*wiremessage.MsgHeader, []byte, error) {
+	err := conn.WriteWireMessage(ctx, wm)
+	if err != nil {
+		labels := []string{NetworkError}
+		if op.Client != nil {
+			op.Client.MarkDirty()
+		}
+		if op.Client != nil && op.Client.TransactionRunning() && !op.Client.Committing {
+			labels = append(labels, TransientTransactionError)
+		}
+		if op.Client != nil && op.Client.Committing {
+			labels = append(labels, UnknownTransactionCommitResult)
+		}
+		return nil, nil, Error{Message: err.Error(), Labels: labels, Wrapped: err}
+	}
+
+	return ReadWireMessageFromConn(ctx, conn, wm)
+}
+
+func (op *Operation) moreToComeRoundTripDirect(ctx context.Context, conn Connection, wm []byte) (*wiremessage.MsgHeader, []byte, error) {
+	err := conn.WriteWireMessage(ctx, wm)
+	if err != nil {
+		if op.Client != nil {
+			op.Client.MarkDirty()
+		}
+		err = Error{Message: err.Error(), Labels: []string{TransientTransactionError, NetworkError}, Wrapped: err}
+	}
+	return nil, bsoncore.BuildDocument(nil, bsoncore.AppendInt32Element(nil, "ok", 1)), err
+}
+
+func (op Operation) processExecuteError(ctx context.Context, err error, operationErr WriteCommandError, setup *execSetup, currIndex int) (error, bool) {
```

Comment:
I just moved the ugly error handling code into a separate function. There's a lot of copy-pasta in here, it's pretty gross